### PR TITLE
Close CUDA 12.0 migration

### PR DESCRIPTION
**Pull request by jakirkham**
Now that a large majority of feedstocks have been rebuilt or have an open migration PR, let's go ahead and close out this PR and add CUDA 12 to the matrix.

This adds CUDA 12.0 & CUDA 12.4. Where CUDA 12.0 represents the oldest CUDA 12 version built against and supported in conda-forge. Similarly CUDA 12.4 represents the latest CUDA version in conda-forge. This way if users are still have packages that depend on an older CUDA 12 version, packages built with CUDA 12.0 will satisfy that need. Also if users are trying to take advantage of packages built with the latest optimizations, they can install packages built with the latest CUDA Toolkit in conda-forge. We keep the latest version pinned to avoid pulling in partially rolled out CUDA Toolkit updates. Meaning we will need to go back through and bump this number when a new CUDA Toolkit becomes available.

xref: https://github.com/conda-forge/conda-forge-pinning-feedstock/issues/5390

<hr>

Checklist
* [ ] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by jakirkham** [comment]
cc @conda-forge/core

File: recipe/conda_build_config.yaml [hunk 1]
```diff
@@ -53,10 +61,14 @@ cuda_compiler:
...
   - nvcc                       # [(linux or win64) and os.environ.get("CF_CUDA_ENABLED", "False") == "True"]
   - nvcc                       # [(linux or win64) and os.environ.get("CF_CUDA_ENABLED", "False") == "True"]
+  - cuda-nvcc                  # [(linux or win64) and os.environ.get("CF_CUDA_ENABLED", "False") == "True"]
+  - cuda-nvcc                  # [(linux or win64) and os.environ.get("CF_CUDA_ENABLED", "False") == "True"]
 cuda_compiler_version:
   - None
   - 11.2                       # [(linux or win64) and os.environ.get("CF_CUDA_ENABLED", "False") == "True"]
   - 11.8                       # [(linux or win64) and os.environ.get("CF_CUDA_ENABLED", "False") == "True"]
+  - 12.0                       # [(linux or win64) and os.environ.get("CF_CUDA_ENABLED", "False") == "True"]
+  - 12.4                       # [(linux or win64) and os.environ.get("CF_CUDA_ENABLED", "False") == "True"]
```
Comment (h-vetinari) [inline recipe/conda_build_config.yaml:71]: Are packages built for 12.0 expected to always successfully build with 12.4? Shouldn't we separate the closing of the 12.0 migration from adding 12.4...?

Comment (xhochy) [reply recipe/conda_build_config.yaml:71]: I remember that in some feedstocks I had to add some code to explicitly use 12.0 to compile. This code should work with 12.4 but as it is not tested, I would feel better if we had a migrator first.

File: recipe/conda_build_config.yaml [hunk 2]
```diff
@@ -159,6 +173,24 @@ docker_image:                                   # [os.environ.get("BUILD_PLATFOR
...
   - quay.io/condaforge/linux-anvil-cuda:11.8              # [aarch64 and os.environ.get("CF_CUDA_ENABLED", "False") == "True" and os.environ.get("BUILD_PLATFORM") == "linux-64"]
 
+  # CUDA 12.0
+  - quay.io/condaforge/linux-anvil-cos7-x86_64     # [linux64 and os.environ.get("BUILD_PLATFORM") == "linux-64" and os.environ.get("CF_CUDA_ENABLED", "False") == "True"]
+  # case: native compilation (build == target)
+  - quay.io/condaforge/linux-anvil-ppc64le         # [ppc64le and os.environ.get("BUILD_PLATFORM") == "linux-ppc64le"]
+  - quay.io/condaforge/linux-anvil-aarch64         # [aarch64 and os.environ.get("BUILD_PLATFORM") == "linux-aarch64"]
+  # case: cross-compilation (build != target)
+  - quay.io/condaforge/linux-anvil-cos7-x86_64     # [ppc64le and os.environ.get("BUILD_PLATFORM") == "linux-64"]
+  - quay.io/condaforge/linux-anvil-cos7-x86_64     # [aarch64 and os.environ.get("BUILD_PLATFORM") == "linux-64"]
```
Comment (h-vetinari) [inline recipe/conda_build_config.yaml:183]: I'm also pretty sure that the selectors here need `os.environ.get("CF_CUDA_ENABLED", "False") == "True"` (which was active in the migrator through the selector on `docker_images`).

**Review by h-vetinari — COMMENTED** [review]
I don't think we should add 12.4 here; I'd be open to dropping 11.2 in the same PR though.

**Review by mariusvniekerk — APPROVED** [review]

**Comment by h-vetinari** [comment]
I think we should move forward here with dropping 11.2 and adding 12.0. Let's split off 12.4.

File: recipe/conda_build_config.yaml [hunk 3]
```diff
@@ -8,6 +8,8 @@ c_compiler_version:            # [unix]
   - 16                         # [osx]
   - 10                         # [os.environ.get("CF_CUDA_ENABLED", "False") == "True" and linux]
   - 11                         # [os.environ.get("CF_CUDA_ENABLED", "False") == "True" and linux]
+  - 12                         # [os.environ.get("cf_cuda_enabled", "false") == "true" and linux]
+  - 12                         # [os.environ.get("cf_cuda_enabled", "false") == "true" and linux]
```
Comment (h-vetinari) [inline recipe/conda_build_config.yaml:12]: This is incorrectly lower-cased here (that environment variable is not set by smithy)

**Review by h-vetinari — CHANGES_REQUESTED** [review]
The announced date for dropping 11.2 has now passed; given that and the various unanswered review requests here, I've opened https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/5799 which builds on this PR and incorporates them.

File: recipe/conda_build_config.yaml [hunk 4]
```diff
@@ -53,10 +61,14 @@ cuda_compiler:
...
+  - cuda-nvcc                  # [(linux or win64) and os.environ.get("CF_CUDA_ENABLED", "False") == "True"]
 cuda_compiler_version:
   - None
   - 11.2                       # [(linux or win64) and os.environ.get("CF_CUDA_ENABLED", "False") == "True"]
   - 11.8                       # [(linux or win64) and os.environ.get("CF_CUDA_ENABLED", "False") == "True"]
+  - 12.0                       # [(linux or win64) and os.environ.get("CF_CUDA_ENABLED", "False") == "True"]
+  - 12.4                       # [(linux or win64) and os.environ.get("CF_CUDA_ENABLED", "False") == "True"]
 cuda_compiler_version_min:
   - None                       # [osx]
   - 11.2                       # [linux or win64]
```
Comment (h-vetinari) [inline recipe/conda_build_config.yaml:74]: This also needs to be bumped.

**Comment by h-vetinari** [comment]
If you plan to rebase this PR now that #5799 is in, you can effectively condense it to a revert of https://github.com/conda-forge/conda-forge-pinning-feedstock/commit/688ab0df767d3046c771a8014577d71db36ae658

**Comment by h-vetinari** [comment]
@jakirkham, what's the status here? What still needs to be done from your POV? (since you asked not to close this migration as part of #5799)

**Comment by h-vetinari** [comment]
> @jakirkham, what's the status here? What still needs to be done from your POV?

Gentle ping on this

**Comment by h-vetinari** [comment]
I've opened a new PR for this in #6263

**Comment by kkraus14** [comment]
Made obsolete by https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/6263

**Pull request closed without merging**